Get values of grouped columns (#1908)
JuliaData/DataFrames.jl: docs/src/lib/indexing.md

If column indexing using `Symbol` names in `cols` is performed, the order of columns in the operation is specified
by the order of names.

## Indexing `GroupedDataFrame`s

> [`GroupedDataFrame`](@ref) implements the dictionary interface and so supports additional methods for indexing:

Reviewer: ```suggestion
[`GroupedDataFrame`](@ref) implements the `AbstractDict` interface and so supports additional methods for indexing:
```

> * `gd[i::Int]` -> Get the `i`th group

Reviewer 2: ```suggestion
* `gd[i::Integer]` -> Get the `i`th group
```

* `gd[key::NamedTuple]` -> Get the group corresponding to the dictionary key `key` (see [`keys(::GroupedDataFrame)`](@ref)).

>   The fields of the `NamedTuple` must be in the same order, which is the order of the columns passed to [`groupby`](@ref).

Reviewer: ```suggestion
  The fields of the `NamedTuple` must match (including order) the grouping columns passed to [`groupby`](@ref).
```

If column indexing using `Symbol` names in `cols` is performed, the order of columns in the operation is specified
by the order of names.

## Indexing `GroupedDataFrame`s

[`GroupedDataFrame`](@ref) implements the `AbstractDict` interface and so supports additional methods for indexing:

* `gd[i::Integer]` -> Get the `i`th group
* `gd[key::GroupKey]` -> Get the group corresponding to the [`GroupKey`](@ref)
  `key` (see [`keys(::GroupedDataFrame)`](@ref)). This should be nearly as fast
  as integer indexing.

> * `gd[key::NamedTuple]` -> Get a group given the values of the grouped columns.

Reviewer 2: ```suggestion
* `gd[key::NamedTuple]` -> Get the group corresponding to the given values of the grouping columns.
```

  The fields of the `NamedTuple` must match the grouping columns columns passed to [`groupby`](@ref) (including order).
* `gd[key::Tuple]` -> Same as previous, but omitting the names on `key`.

> * `get(gd, key, default)` -> Get group by key, returning `default` if it does not exist.

Reviewer 2: ```suggestion
* `get(gd, key, default)` -> Get group by `key`, returning `default` if it does not exist.
```

If column indexing using `Symbol` names in `cols` is performed, the order of columns in the operation is specified
by the order of names.

## Indexing `GroupedDataFrame`s

A [`GroupedDataFrame`](@ref) can behave as either an `AbstractVector` or `AbstractDict` depending on the type
of index used. Integers (or arrays of them) trigger vector-like indexing while `Tuples`s and `NamedTuple`s
trigger dictionary-like indexing. An intermediate between the two is the [`GroupKey`](@ref) type
returned by [`keys(::GroupedDataFrame)`](@ref), which behaves similarly to a `NamedTuple` but has
performance on par with integer indexing.

> The "elements" or "values" of a `GroupedDataFrame` are [`SubDataFrame`](@ref)s of its parent.

Reviewer: ```suggestion
The elements of a `GroupedDataFrame` are [`SubDataFrame`](@ref)s of its parent.
```

* `gd[i::Integer]` -> Get the `i`th group.

> * `gd[a::AbstractArray{<:Integer}]` -> Get a reduced `GroupedDataFrame` containing only the groups

Reviewer 2: This should be a correct contract - right? If yes - can we have all these options in the tests. Thank you!
```suggestion
* `gd[a::Union{AbstractVector{<:Integer}, Colon, Not}]` -> Get a reduced `GroupedDataFrame` containing only the groups
```

Author: This behavior hasn't changed in this PR (aside from restricting the signature), so I'd rather update this in a follow-up PR if that's alright.

Reviewer 2: OK

  with indices in `a`.
* `gd[key::NamedTuple]` -> Get the group corresponding to the given values of the
  grouping columns. The fields of the `NamedTuple` must match the grouping columns
  columns passed to [`groupby`](@ref) (including order).
* `gd[key::Tuple]` -> Same as previous, but omitting the names on `key`.

> * `get(gd, key, default)` -> Get group by (named) Tuple `key`, returning `default` if it does not exist.

Reviewer 2: I think it is clearer to be explicit
```suggestion
* `get(gd, key, default)` -> Get group by `NamedTuple` or `Tuple` `key`, returning `default` if it does not exist.
```
